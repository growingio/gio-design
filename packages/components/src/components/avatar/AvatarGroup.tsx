import React from 'react';
import classNames from 'classnames';
import _ from 'lodash';
import Avatar from './Avatar';
import { AvatarGroupProps, UserAvatarType } from './interfaces';
import usePrefixCls from '../../utils/hooks/use-prefix-cls';

const AvatarGroup: React.FC<AvatarGroupProps> = (props: AvatarGroupProps) => {
  const { number = 5, users = [], className, placement = 'bottom', style, displayTooltip = true } = props;
  const prefixCls = usePrefixCls('avatar');

  let children = null;
  const renderAvatarGroup = (sliceUsers: UserAvatarType[]) =>
    sliceUsers.map((user) => (
      <Avatar key={user.name} displayTooltip={displayTooltip} placement={placement} {...user}>
        {user.name}
      </Avatar>
    ));
  const renderAvatarRest = (restUsers: UserAvatarType[]) => (
    <Avatar className={`${prefixCls}-rest`} omit={false}>{`+${restUsers.length}`}</Avatar>
  );
  const classString = classNames(className, `${prefixCls}-group`);

  if (users.length === 0) {
    return null;
  }
  if (users.length <= number) {
    children = renderAvatarGroup(users);
  } else {
    const sliceUsers = _.slice(users, 0, number - 1);
    const restUsers = _.slice(users, number - 1, users.length);
    children = (
      <>
        {renderAvatarGroup(sliceUsers)}
        {renderAvatarRest(restUsers)}
      </>
    );
  }

  return (
    <div className={classString} style={style}>
      {children}
    </div>
  );
};
export default AvatarGroup;
