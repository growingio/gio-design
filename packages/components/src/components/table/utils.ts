import { clone, has, get, join, set } from 'lodash';
import { ColumnsType, InnerColumnsType } from './interface';

export const translateInnerColumns = <RecordType>(columns: ColumnsType<RecordType>): InnerColumnsType<RecordType> =>
  clone(columns).map((cloneColumn) => {
    if (!has(cloneColumn, 'key')) {
      if (has(cloneColumn, 'dataIndex')) {
        if (Array.isArray(get(cloneColumn, 'dataIndex'))) {
          cloneColumn.key = join(get(cloneColumn, 'dataIndex'), '-');
        } else {
          cloneColumn.key = get(cloneColumn, 'dataIndex');
        }
      } else {
        console.warn('gio-design table: column key or dataIndex must have one');
      }
    }
    if (has(cloneColumn, 'children')) {
      set(cloneColumn, 'children', translateInnerColumns(get(cloneColumn, 'children')));
    }
    return cloneColumn;
  }) as InnerColumnsType<RecordType>;

export default {}
