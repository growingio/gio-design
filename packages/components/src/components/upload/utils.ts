import { defaultRootPrefixCls } from '../config-provider';
import { IRcFile, IUploadFile, STATUS_NOT_YET } from './interface';

export const getUid = (): string => `${defaultRootPrefixCls}-upload-${Date.now()}`;

export const fileToObject = (file: IRcFile): IUploadFile => ({
  ...file,
  lastModified: file.lastModified,
  lastModifiedDate: file.lastModifiedDate,
  name: file.name,
  size: file.size,
  type: file.type,
  uid: file.uid,
  percent: 0,
  originFile: file,
});

export const getEmptyFileObj = (): IUploadFile => ({
  uid: getUid(),
  size: 0,
  name: '本地上传',
  type: '$empty-file',
  status: STATUS_NOT_YET,
});

export const requestImage = (src: string, cors?: 'anonymous' | 'use-credentials'): Promise<HTMLImageElement> =>
  new Promise((resolve, reject) => {
    const img = new Image();

    img.addEventListener('load', () => resolve(img));

    img.addEventListener('error', (err) => reject(err));

    if (cors) {
      img.setAttribute('crossOrigin', cors);
    }
    img.src = src;
  });
/**
 *
 * @param {File | Blob} file 图片文件
 * @returns {Promise<string> } 返回值是图片的 base64 字符串
 */
export const imageFile2DataUrl = (file: File | Blob): Promise<string> =>
  new Promise((resolve, reject) => {
    const canvas = document.createElement('canvas');
    const url = window.URL.createObjectURL(file);
    requestImage(url)
      .then((img) => {
        const { width, height } = img;
        canvas.width = width;
        canvas.height = height;

        const ctx = canvas.getContext('2d');
        ctx?.drawImage(img, 0, 0, width, height);
        const dataUrl = canvas.toDataURL();
        resolve(dataUrl);
      })
      .catch(reject);
  });

/**
 * 图片 base64 转为 File 格式
 */
export const dataUrl2ImageFile = (dataUrl = ''): File => {
  const arr = dataUrl.split(',');
  const mime = arr[0].match(/:(.*?);/)?.[1];
  const bstr = atob(arr[1] ?? '');
  let n = bstr.length;
  const u8arr = new Uint8Array(n);
  while (n) {
    n -= 1;
    u8arr[n] = bstr.charCodeAt(n);
  }
  return new File([u8arr], 'web-image', { type: mime });
};

/**
 *  网络 url 图片获取
 * @param {string} url
 */
export const fetchImageFileFromUrl = (
  url: string
): Promise<{
  originFile: IRcFile;
  dataUrl: string;
}> =>
  new Promise((resolve, reject) => {
    const canvas = document.createElement('canvas');
    requestImage(url, 'anonymous')
      .then((img) => {
        const { width, height } = img;
        canvas.width = width;
        canvas.height = height;

        const ctx = canvas.getContext('2d');
        ctx?.drawImage(img, 0, 0, width, height);
        let dataUrl = '';
        try {
          dataUrl = canvas.toDataURL();
        } catch (error) {
          reject(error);
        }
        const file = dataUrl2ImageFile(dataUrl);
        (file as IRcFile).uid = getUid();
        resolve({
          originFile: file as IRcFile,
          dataUrl,
        });
      })
      .catch((err) => reject(err));
  });

export const isOnlyAcceptImg = (accept?: string): boolean =>
  accept
    ? accept
        .split(',')
        .map((_) => _.trim())
        .filter((_) => !/((image\/)|\.)(dwg|dxf|gif|jp2|jpeg|jpg|jpe|png|svf|tif|tiff)|image\/\*/.test(_)).length === 0
    : false;

export const isImageFile = (file: IUploadFile): boolean => file.type.startsWith('image/');

/**
 * 暂时没有用起来，文件头在处理 docx/xlsx 等微软文件时候的区分方法暂时没写
 * @param file
 */
export const getFileType = (file: File | Blob): Promise<string | undefined> =>
  new Promise((resolve) => {
    const fr = new FileReader();

    fr.addEventListener('load', (e: any) => {
      const view = new DataView(e.target.result);
      const first4Byte = view.getUint32(0, false);
      const hexValue = Number(first4Byte).toString(16).toUpperCase();

      switch (hexValue) {
        case 'FFD8FFE0':
        case 'FFD8FFE1':
        case 'FFD8FFE2':
        case 'FFD8FFE3':
          resolve('JPEG/JPG');
          break;
        default:
          resolve(undefined);
          break;
      }
    });

    fr.readAsArrayBuffer(file);
  });
