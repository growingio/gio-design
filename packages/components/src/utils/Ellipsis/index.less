.gio-ui-ellipsis {
  display: inline-block;
  display: -webkit-box;
  overflow: hidden;
  text-overflow: ellipsis;
  word-break: break-all;
  -webkit-line-clamp: 2;
  &__line--2 {
    -webkit-line-clamp: 2;
  }
  &__line--1 {
    -webkit-line-clamp: 1;
  }
}
.gio-ui-single-line-ellipsis {
  display: inline-block;
  width: 100%;
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
  word-wrap: normal;
}
