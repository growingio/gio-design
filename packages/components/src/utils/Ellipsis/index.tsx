import * as React from 'react';
import { Tooltip } from 'antd';

interface P {
  line?: number;
  title?: React.ReactNode;
  maxWidth?: string;
  style?: object;
  placement?:
    | 'top'
    | 'left'
    | 'right'
    | 'bottom'
    | 'topLeft'
    | 'topRight'
    | 'bottomLeft'
    | 'bottomRight'
    | 'leftTop'
    | 'leftBottom'
    | 'rightTop'
    | 'rightBottom';
}

class Ellipsis extends React.PureComponent<P> {
  public state = {
    tooltip: false,
  };
  private textRef: HTMLSpanElement | null = null;
  public componentDidMount() {
    this.detectTooltip();
  }
  public componentDidUpdate(prevProps: P) {
    if (prevProps.title !== this.props.title) {
      this.detectTooltip();
    }
  }
  public render() {
    const { children, line, title, maxWidth, style, placement } = this.props;
    const _style = { maxWidth, ...style };
    if (!line) {
      return (
        <Tooltip
          title={this.state.tooltip ? title : ''}
          placement={placement}
          overlayStyle={{ wordWrap: 'break-word' }}
        >
          <span className='gio-ui-single-line-ellipsis' ref={this.setTextRef} style={_style}>
            {children}
          </span>
        </Tooltip>
      );
    }
    return (
      <Tooltip title=''>
        <span
          className={`gio-ui-ellipsis gio-ui-ellipsis__line--${line}`}
          style={{
            boxOrient: 'vertical',
            lineClamp: line,
            ..._style,
          }}
          ref={this.setTextRef}
        >
          {children}
        </span>
      </Tooltip>
    );
  }
  private setTextRef = (ref: HTMLSpanElement) => {
    if (ref) {
      this.textRef = ref;
    } else {
      this.textRef = null;
    }
  };
  private detectTooltip = () => {
    this.setTooltip();
    // IE、Firefox下多行省略不兼容，限制行数。
    this.limitLineNumbers();
  };
  private setTooltip = () => {
    let tooltip = false;
    if (this.textRef && this.textRef.scrollWidth > this.textRef.clientWidth) {
      tooltip = true;
    }
    if (tooltip !== this.state.tooltip) {
      this.setState({ tooltip });
    }
  };
  private limitLineNumbers = () => {
    const { line } = this.props;
    if (!line) {
      return;
    }
    if (this.textRef) {
      const lineHeight = parseInt(getComputedStyle(this.textRef)['line-height' as any], 10);
      if (typeof lineHeight === 'number' && !isNaN(lineHeight)) {
        this.textRef.style.maxHeight = line * lineHeight + 'px';
      }
    }
  };
}

export default Ellipsis;
