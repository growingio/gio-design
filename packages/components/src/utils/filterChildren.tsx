import React from 'react';

export type IFilterChildren = (
  children: React.ReactNode,
  filter: (child: React.ReactNode) => boolean
) => React.ReactNodeArray;

const filterChildren: IFilterChildren = (children, filter) => {
  const result: React.ReactNodeArray = [];

  React.Children.forEach(children, (child) => {
    if (filter(child)) {
      result.push(child);
    }
  });

  return result;
};

export default filterChildren;
