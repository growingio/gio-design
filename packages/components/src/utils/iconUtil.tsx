import * as React from 'react';
import CaretDownOutlined from '@gio-design/icons/es/CaretDownOutlined';
import LoadingOutlined from '@gio-design/icons/es/LoadingOutlined';
import CheckOutlined from '@gio-design/icons/es/CheckOutlined';
import Close from '@gio-design/icons/es/Close';
import CloseCircleFilled from '@gio-design/icons/es/CloseCircleFilled';
import SearchOutlined from '@gio-design/icons/es/SearchOutlined';

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export default function getIcons({
  suffixIcon,
  clearIcon,
  menuItemSelectedIcon,
  removeIcon,
  loading,
  multiple,
  prefixCls,
}: {
  suffixIcon?: React.ReactNode;
  clearIcon?: React.ReactNode;
  menuItemSelectedIcon?: React.ReactNode;
  removeIcon?: React.ReactNode;
  loading?: boolean;
  multiple?: boolean;
  prefixCls: string;
}) {
  // Clear Icon
  let mergedClearIcon = clearIcon;
  if (!clearIcon) {
    mergedClearIcon = <CloseCircleFilled size='12px' />;
  }

  // Arrow item icon
  let mergedSuffixIcon = null;
  if (suffixIcon !== undefined) {
    mergedSuffixIcon = suffixIcon;
  } else if (loading) {
    mergedSuffixIcon = <LoadingOutlined rotating size='12px' />;
  } else {
    const iconCls = `${prefixCls}-suffix`;
    mergedSuffixIcon = ({ open, showSearch }: { open: boolean; showSearch: boolean }) => {
      if (open && showSearch) {
        return <SearchOutlined className={iconCls} size='12px' />;
      }
      return <CaretDownOutlined className={iconCls} size='12px' />;
    };
  }

  // Checked item icon
  let mergedItemIcon = null;
  if (menuItemSelectedIcon !== undefined) {
    mergedItemIcon = menuItemSelectedIcon;
  } else if (multiple) {
    mergedItemIcon = <CheckOutlined size='12px' />;
  } else {
    mergedItemIcon = null;
  }

  let mergedRemoveIcon = null;
  if (removeIcon !== undefined) {
    mergedRemoveIcon = removeIcon;
  } else {
    mergedRemoveIcon = <Close size='12px' />;
  }

  return {
    clearIcon: mergedClearIcon,
    suffixIcon: mergedSuffixIcon,
    itemIcon: mergedItemIcon,
    removeIcon: mergedRemoveIcon,
  };
}
