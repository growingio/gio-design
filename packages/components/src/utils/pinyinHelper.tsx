/* eslint-disable no-param-reassign */
/* eslint-disable @typescript-eslint/no-require-imports */
const pinyinMatch = require('pinyin-match');

export default function isContain(target = '', source = ''): boolean {
  if (!target) {
    target = '';
  }
  if (!source) {
    source = '';
  }
  return !!pinyinMatch.match(target, source);
}
