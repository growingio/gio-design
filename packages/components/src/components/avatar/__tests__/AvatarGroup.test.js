import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import AvatarGroup from '../AvatarGroup';
import '../../../../es/components/avatar/style/index.css';
import { waitForComponentToPaint } from '../../../utils/test';
import image from './icon.jpeg';

describe('Testing AvatarGroup', () => {
  const users = [
    {
      name: 'li',
      src: image,
    },
    {
      name: 'pan',
    },
    {
      name: 'leng',
      src: image,
    },
    {
      name: 'liu',
    },
    {
      name: 'wang',
      src: image,
    },
    {
      name: 'tong',
      src: image,
    },
  ];

  it('should be stable', () => {
    const wrapper = renderer.create(<AvatarGroup users={users} />).toJSON();
    expect(wrapper).toMatchSnapshot();
  });

  it('should be render rightly', () => {
    let wrapper = mount(<AvatarGroup users={users} />);
    expect(wrapper.exists('.gio-avatar-group')).toBe(true);
    wrapper = wrapper.find('.gio-avatar');
    expect(wrapper.at(0).childAt(0).type()).toBe('img');
    expect(wrapper.at(1).childAt(0).type()).toBe('span');
    expect(wrapper.at(1).childAt(0).text()).toBe('p');
    expect(wrapper.at(2).childAt(0).type()).toBe('img');
    expect(wrapper.at(3).childAt(0).type()).toBe('span');
    expect(wrapper.at(3).childAt(0).text()).toBe('l');
    expect(wrapper.at(4).childAt(0).text()).toBe(`+${users.length - 4}`);
  });

  it('should have two character avatars', () => {
    const wrapper = mount(<AvatarGroup users={users} />);
    expect(
      wrapper
        .childAt(0)
        .children()
        .filterWhere((n) => n.find('.gio-avatar').childAt(0).type() === 'img')
    ).toHaveLength(2);
  });

  test('props number', () => {
    const wrapper = mount(<AvatarGroup users={users} number={4} />).childAt(0);
    expect(wrapper.find('.gio-avatar')).toHaveLength(4);
    expect(wrapper.last().exists('.gio-avatar-rest')).toBe(true);
  });

  test('props placement', (done) => {
    const wrapper = mount(<AvatarGroup users={users} number={4} />);
    wrapper.setProps({ placement: 'top' });
    wrapper.find('.gio-avatar').at(0).simulate('mouseenter');
    waitForComponentToPaint(wrapper).then(() => {
      expect(wrapper.exists('.gio-tooltip-placement-top')).toBe(true);
      done();
    });
  });

  test('props users', () => {
    expect(mount(<AvatarGroup users={[]} number={4} />).exists('.gio-avatar-group')).toBe(false);
    expect(mount(<AvatarGroup users={users.slice(0, 3)} number={4} />).find('.gio-avatar')).toHaveLength(3);
  });
});
