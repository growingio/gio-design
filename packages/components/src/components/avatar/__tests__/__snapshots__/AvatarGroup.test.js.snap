// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing AvatarGroup should be stable 1`] = `
<div
  className="gio-avatar-group"
>
  <span
    className="gio-avatar gio-avatar-df"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <img
      alt="avatar"
      onError={[Function]}
      src="test-file-stub"
    />
  </span>
  <span
    className="gio-avatar gio-avatar-df"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <span
      style={
        Object {
          "transform": "scale(1)",
        }
      }
    >
      p
    </span>
  </span>
  <span
    className="gio-avatar gio-avatar-df"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <img
      alt="avatar"
      onError={[Function]}
      src="test-file-stub"
    />
  </span>
  <span
    className="gio-avatar gio-avatar-df"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <span
      style={
        Object {
          "transform": "scale(1)",
        }
      }
    >
      l
    </span>
  </span>
  <span
    className="gio-avatar-rest gio-avatar gio-avatar-df"
  >
    <span
      style={
        Object {
          "transform": "scale(1)",
        }
      }
    >
      +2
    </span>
  </span>
</div>
`;
